File: linked-lists/remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.go
package removenthnodefromendoflist

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{Val: 0, Next: head}
	first := dummy
	second := dummy

	for n >= 0 {
		n--
		first = first.Next
	}

	for first != nil {
		first = first.Next
		second = second.Next
	}
	second.Next = second.Next.Next
	return dummy.Next
}

File: linked-lists/remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.ts
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
  const dummy: ListNode | null = new ListNode(0, head!);
  let first: ListNode | null = dummy;
  let second: ListNode | null = dummy;

  while (n >= 0) {
    n--;
    first = first!.next;
  }

  while (first !== null) {
    first = first.next;
    second = second!.next;
  }

  second!.next = second!.next!.next;
  return dummy.next;
}
